### PR TITLE
Thumbnail: thumbfast integration

### LINKED ISSUE #43: Thumbnails
There is a thumbnail script out: https://github.com/TheAMM/mpv_thumbnail_script

From an earlier discussion in some IRC channel, I thought you'd like to take a look into it for how much work it might be to add this here. Currently, it replaces the OSC with a custom one since it needs to accurately determine the hover regions of the hover bars, or something. 

```
[12-06][00:17:47] <FichteFoll> so, I assume this need to be added to torque-progressbar manually?
[12-06][00:18:53] <@AMM> I looked at it
[12-06][00:19:09] <@AMM> It'll be a fucking pain to integrate it properly
```

### PR DESCRIPTION
**Pull request by po5**
Closes #43.  
This adds support for [thumbfast](https://github.com/po5/thumbfast).

Did this over a month ago but finally making a PR for it.  
I don't expect this to be merged as-is, as I'm sure my unfamiliarity with the codebase led to bad choices. But it works.

Hope you can guide me to clean this up.

Preview

[thumbfast-progressbar.webm](https://user-images.githubusercontent.com/42466980/198861924-fee3717b-9bae-42c0-82b4-624eb0d27d7f.webm)

### CODE REVIEW

**Comment by torque** [comment]
Hey, thank you for this PR, this is a feature I long wanted to implement but never got to. I'll take a look at this soon.

**Comment by torque** [comment]
The only real functional issue I ran into while playing around with this was that it was behaving incorrectly on high-DPI displays. This is because `progressbar` does some internal management of the OSD on high-dpi displays and `thumbfast` does as well, so essentially the positioning of the thumbnail was getting scaled twice. I ended up going down a bit of a rabbit hole in fixing this (including finally addressing the extremely long-standing issue of `progressbar` not automatically updating its scaling when switching between displays with different DPIs, so thanks for the motivation to fix that). Essentially the fix was simple, I expose the non-scaled window/mouse coordinates to the `thumbfast` call now, but it required touching various parts.

Since that fix was a bit more invasive, I went ahead and committed those changes directly to your PR. I hope you don't mind. I did end up making minor changes to the `Thumbnail` class you added, primarily:

1.  I cleaned out some boilerplate (mostly the animation stuff that the integration wasn't using anyway, but that I had sort of implicitly baked into the object inheritance stupidly
2. I adjusted some minor lua-isms to their more moonscripty counterparts (e.g. `~=` to `!=`)
3. I updated the way the `Thumbnail` object is created. Since the thumbnail information comes in asynchronously in a callback, it's a little goofy to deal with, but it seems to work ok in practice
4. I added an option to overall enable/disable thumbnails in line with the way the other UI widgets can be disabled

Thanks for submitting this PR and for developing thumbfast, it's really nice to have this functionality. With these changes, I'm ready to merge this, but if you're interested in doing any testing on your end, let me know and I'll hold off for a bit.

File: src/Thumbnail.moon [hunk 1]
```diff
@@ -4,58 +4,40 @@ class Thumbnail extends BarAccent
...
+			mp.commandv( 'script-message-to', 'thumbfast', 'clear' )
+			@needsUpdate = true
 
 	redraw: =>
 		if @active
 			super!
-			if not thumbfast.disabled and thumbfast.width ~= 0 and thumbfast.height ~= 0
-				hoverTime = mp.get_property_number( 'duration', 0 )*Mouse.x/Window.w
-				mp.commandv( 'script-message-to', 'thumbfast', 'thumb',
+			if Mouse.x != @lastX and not @thumbfast.disabled and @thumbfast.width >= 0 and @thumbfast.height >= 0
```
Comment (po5) [inline src/Thumbnail.moon:29]: Should probably be `@thumbfast.width > 0 and @thumbfast.height > 0`.

**Comment by po5** [comment]
Thanks for the hidpi stuff, I don't have such a display and wouldn't know how to fake it for testing.

The last commit 7297ab0f61aabc6dfda2c9888a767b874f73d035 broke the timeline for me. Top title still shows up fine.  
Sorry for the awful quality. Previous commit works.

[mpvprogressbar.webm](https://user-images.githubusercontent.com/42466980/200133323-e2735841-3c17-4718-9653-37a8d8966c03.webm)

Not sure which info you need for troubleshooting, if it works for you.  
(Close to) latest mpv built at this commit https://github.com/mpv-player/mpv/commit/6acb7db9facd41f67aebb54a0daeddaaa30d54b3, linux.

Currently most integrations draw a border around the thumbnail, to make it fit in better than just a floating image.  
Is this something you would want to have?

File: src/Thumbnail.moon (hunk 1 above)
Comment (torque) [reply src/Thumbnail.moon:29]: cool, that's a nice simplification

**Comment by torque** [comment]
> Did some digging, it's because we can receive multiple `thumbfast-info` calls (when swiching videos, or if the aspect ratio changes mid-video). Adding more than one `Thumbnail` element is what breaks it. Keep in mind we do need to update our thumbfast object to position everything correctly. Can't just ignore all calls after the first.

Hmm, yeah I was worried about something like this happening, but I actually did not encounter problems when testing with it. I did some simple checking with a hybrid playlist of one network video and one local video, but I didn't check multiple local videos in a row. Thanks for figuring out the cause, it should be simple to make it so that successive invocation of the `thumbfast-info` call updates the existing object rather than trying to create a new one.

> Currently most integrations draw a border around the thumbnail, to make it fit in better than just a floating image. Is this something you would want to have?

Yeah, I had this in mind as I tweaked the `Thumbnail` class to make it simpler to add a drawing underneath it, but I didn't actually end up implementing any drawing. I honestly don't think that the bare thumbnail feels terribly out of place with the general minimalism of this script, though there are obviously cases where having a border to differentiate it from the background would be good (e.g. a video/scene that is a flat color). I mentally put that in the "to do later" bucket along with experimenting with animating the thumbnail into place (which probably won't look great for layering reasons, among others). If it's something you'd like to work on though, I'd be happy to provide some guidance.

**Comment by po5** [comment]
It wouldn't draw the thumbnail when switching files with the cursor over the timeline but not moving.

Looks good to merge now.

**Comment by torque** [comment]
Thanks again for the contribution!

**Pull request merged**